### PR TITLE
feat: support `FormData`

### LINKED ISSUE #516: Testing file upload without form-data library
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the issue has not already been raised

### Issue

Hello there.

I would like to add testing to my endpoint which uses `fastify/multipart` to accept a file upload and then writes it to the filesystem.

In the testing directory of this repository I saw that popular `form-data` library is used, however, I think this can also be achieved without it?

What I have come up with, but for some reason does not work:
```ts
const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const filePath = join(__dirname, '../example.jpg')

describe('POST /asset', () => {
        it('should return 201 when the payload is valid', async () => {
            const rawData = readFileSync(filePath, 'binary')

            const body = new FormData()
            const blob = new Blob([rawData])

            body.set('file', blob, 'example.jpg')

            const res = await app.inject({
                method: 'POST',
                url: '/asset',
                payload: {
                    file: body.get('file')
                }
            })

            assert.deepStrictEqual(res.statusCode, 201)
        })
    })
 ```
    
However, this gives me the following error:
```
{"statusCode":400,"code":"FST_ERR_VALIDATION","error":"Bad Request","message":"body/file should be a file"}
```

This is my route definition (I do not think this is relevant, but maybe it could be):
```ts
fastify.post(
        '/',
        {
            schema: {
                consumes: ['multipart/form-data'],
                body: {
                    type: 'object',
                    required: ['file'],
                    properties: {
                        file: { isFile: true }
                    }
                }
            }
        },
        createAsset
    )
```
    
Is there any way to make this testing work?

Thank you very much in advance.

### LINKED ISSUE #1017: Testing fastify/multipart file upload without form-data library
Hello there.

I would like to add testing to my endpoint which uses `fastify/multipart` to accept a file upload and then writes it to the filesystem.

In the testing directory of this repository I saw that popular `form-data` library is used, however, I think this can also be achieved without it?

What I have come up with, but for some reason does not work:
```ts
const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const filePath = join(__dirname, '../example.jpg')

describe('POST /asset', () => {
        it('should return 201 when the payload is valid', async () => {
            const rawData = readFileSync(filePath, 'binary')

            const body = new FormData()
            const blob = new Blob([rawData])

            body.set('file', blob, 'example.jpg')

            const res = await app.inject({
                method: 'POST',
                url: '/asset',
                payload: {
                    file: body.get('file')
                }
            })

            assert.deepStrictEqual(res.statusCode, 201)
        })
    })
 ```
    
However, this gives me the following error:
```
{"statusCode":400,"code":"FST_ERR_VALIDATION","error":"Bad Request","message":"body/file should be a file"}
```

This is my route definition (I do not think this is relevant, but maybe it could be):
```ts
fastify.post(
        '/',
        {
            schema: {
                consumes: ['multipart/form-data'],
                body: {
                    type: 'object',
                    required: ['file'],
                    properties: {
                        file: { isFile: true }
                    }
                }
            }
        },
        createAsset
    )
```
    
Is there any way to make this testing work?

Thank you very much in advance.

### PR DESCRIPTION
**Pull request by climba03003**
Support FormData and provide the `Content-Type` header automatically.

Closes https://github.com/fastify/fastify-multipart/issues/516
Closes https://github.com/fastify/help/issues/1017

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by climba03003** [comment]
The coverage is really broken in supporting `FormData`.
It can only be used in `node > 18` but we still supporting `14 <= node < 18`.

File: lib/form-data.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    typeof payload.delete === 'function' &&
+    typeof payload.get === 'function' &&
+    typeof payload.getAll === 'function' &&
+    typeof payload.has === 'function' &&
+    typeof payload.set === 'function' &&
+    payload[Symbol.toStringTag] === 'FormData'
+  )
+}
+
+function formDataToStream (formdata) {
```
Comment (climba03003) [inline lib/form-data.js:24]:
```suggestion
/*
  partial code extraction and refactoring of `undici`.
  MIT License. https://github.com/nodejs/undici/blob/043d8f1a89f606b1db259fc71f4c9bc8eb2aa1e6/lib/web/fetch/LICENSE
  Reference https://github.com/nodejs/undici/blob/043d8f1a89f606b1db259fc71f4c9bc8eb2aa1e6/lib/web/fetch/body.js#L102-L168
*/
function formDataToStream (formdata) {
```

File: test/index.test.js [hunk 2]
```diff
@@ -2012,3 +2012,96 @@ test('request that is destroyed does not error', (t) => {
     t.equal(res.payload, 'hi')
   })
 })
+
+test('undici / native form-data should be handled correctly', (t) => {
+  if (parseInt(process.versions.node.split('.', 1)[0], 10) < 18) {
```
Comment (mcollina) [inline test/index.test.js:2017]: please use the native skip functionality

**Review by mcollina — CHANGES_REQUESTED** [review]
Amazing work! Can you add a test for Node v14 and v16?

File: test/index.test.js [hunk 3]
```diff
@@ -2012,3 +2012,182 @@ test('request that is destroyed does not error', (t) => {
...
+      }
+      i++
+    })
+  })
+}, { skip: globalThis.FormData == null || globalThis.Blob == null })
+
+// used for testing node@14 / node@16
+test('undici form-data should be handled correctly', (t) => {
+  // inline require to prevent global variable pollution
+  const { FormData } = require('undici')
```
Comment (climba03003) [inline test/index.test.js:2106]: I see a way to test in `node@14` and `node@16` now.
It requires to mix `undici` and `node:buffer` to support both `Blob` and `FormData` in old release line.

File: test/index.test.js [hunk 4]
```diff
@@ -2012,3 +2012,99 @@ test('request that is destroyed does not error', (t) => {
...
+    })
+  }, { skip: FormData == null || Blob == null })
+}
+
+// supports >= node@18
+runFormDataUnitTest('native', { FormData: globalThis.FormData, Blob: globalThis.Blob })
+// supports >= node@16
+runFormDataUnitTest('undici', { FormData: require('undici').FormData, Blob: require('node:buffer').Blob })
+// supports >= node@14
+runFormDataUnitTest('formdata-node', { FormData: require('formdata-node').FormData, Blob: require('formdata-node').Blob })
```
Comment (climba03003) [inline test/index.test.js:2110]: Coverage should be fine for all supported node version.
`undici` only supports `FormData` higher than `node@16`.
So, we need `formdata-node` for `node@14`.

It also show that we are not limited to `undici` implementation only, but support all spec compliance `FormData`

**Review by mcollina — APPROVED** [review]
lgtm

**Review by metcoder95 — APPROVED** [review]

**Pull request merged**